### ISSUE TITLE
Segmentation fault while handling trees

### ISSUE THREAD

**Issue by ageron** [roc-lang/roc#7098]
The following app causes a segmentation fault on my Macbook. I'm using this `roc version`:
```
roc nightly pre-release, built from commit 3215a8f on Wed Sep 18 09:02:06 UTC 2024
```

This program builds two trees and checks whether or not they are the same, ignoring the order of the children.

```roc
app [main] {
    pf: platform "https://github.com/roc-lang/basic-cli/releases/download/0.15.0/SlwdbJ-3GR7uBWQo6zlmYWNYOxnvo8r6YABXD-45UOw.tar.br",
}

import pf.Stdout

Tree : [Empty, Node { label : Str, children : List Tree }]

## Are tree1 and tree2 identical, ignoring the order of the children?
isEquivalentTo : Tree, Tree -> Bool
isEquivalentTo = \tree1, tree2 ->
    when (tree1, tree2) is
        (Empty, Empty) -> Bool.true
        (Empty, Node _) -> Bool.false
        (Node _, Empty) -> Bool.false
        (Node node1, Node node2) ->
            if node1.label != node2.label then
                Bool.false
                else

            help : List Tree, List Tree -> Bool
            help = \children1, children2 ->
                when children1 is
                    [] -> children2 == []
                    [child1, .. as rest1] ->
                        maybeIndex =
                            children2
                            |> List.findFirstIndex \child2 ->
                                child2 |> isEquivalentTo child1
                        when maybeIndex is
                            Ok index -> help rest1 (children2 |> List.dropAt index)
                            Err NotFound -> Bool.false
            help [tree1] [tree2]

main =
    tree1 =
        Node {
            label: "Alice",
            children: [
                Node { label: "Bob", children: [] },
                Node { label: "Charlie", children: [] },
            ],
        }
    tree2 =
        Node {
            label: "Alice",
            children: [
                Node { label: "Charlie", children: [] },
                Node { label: "Bob", children: [] },
            ],
        }
    if tree1 |> isEquivalentTo tree2 then
        Stdout.line! "Trees are equivalent"
    else
        Stdout.line! "Trees are different"
```

**Comment by bhansconnect** [comment]
To also put the info here, this is just a stack overflow. I'm not sure why many stack overflows in roc end up being segfaults, but that is what this is:
Need to update:
```elm
            help [tree1] [tree2]
```
to:
```elm
            help node1.children node2.children
```

**Comment by ageron** [comment]
Thanks @bhansconnect . With this fix & the fix for #7065, this code works fine. 👍
